--- emqx	2022-08-31 07:20:47.000000000 -0400
+++ emqx	2022-09-06 19:04:45.000000000 -0400
@@ -390,7 +390,7 @@
 
     ## Merge cuttlefish generated *.args into the vm.args
     TMP_ARG_FILE="$RUNNER_DATA_DIR/configs/vm.args.tmp"
-    cp "$RUNNER_ETC_DIR/vm.args" "$TMP_ARG_FILE"
+    cat "$RUNNER_ETC_DIR/vm.args" > "$TMP_ARG_FILE"
     echo "" >> "$TMP_ARG_FILE"
     echo "-pa \"${REL_DIR}/consolidated\"" >> "$TMP_ARG_FILE"
     sed '/^#/d' "$CUTTLE_GEN_ARG_FILE" | sed '/^$/d' | while IFS='' read -r ARG_LINE || [ -n "$ARG_LINE" ]; do
