--- emqx.cmd	2022-02-21 13:07:23.000000000 -0500
+++ emqx.cmd	2022-02-21 13:11:23.000000000 -0500
@@ -67,7 +67,7 @@
 )
 
 :: Write the erl.ini file to set up paths relative to this script
-@call :write_ini
+:: @call :write_ini
 
 :: If a start.boot file is not present, copy one from the named .boot file
 @if not exist "%rel_dir%\start.boot" (
@@ -255,7 +255,7 @@
 
 :: Attach to a running node
 :attach
-:: @start "%node_name% attach" 
+:: @start "%node_name% attach"
 @start "%node_name% attach" %werl% -boot "%clean_boot_script%" ^
   -remsh %node_name% %node_type% console_%node_name% -setcookie %node_cookie%
 @goto :eof
