--- emqx.cmd	2022-02-21 16:30:22.000000000 -0500
+++ emqx.cmd	2022-02-21 16:43:52.000000000 -0500
@@ -30,12 +30,19 @@
 @for %%A in ("%script_dir%\..") do @(
   set rel_root_dir=%%~fA
 )
+@if exist %rel_root_dir% (
+  cd /d %rel_root_dir%
+)
 @set rel_dir=%rel_root_dir%\releases\%rel_vsn%
 @set RUNNER_ROOT_DIR=%rel_root_dir%
 
 @set etc_dir=%rel_root_dir%\etc
 @set lib_dir=%rel_root_dir%\lib
-@set data_dir=%rel_root_dir%\data
+@if defined EMQX_NODE__DATA_DIR (
+  set data_dir=%EMQX_NODE__DATA_DIR%
+) else (
+  @set data_dir=%rel_root_dir%\data
+)
 @set emqx_conf=%etc_dir%\emqx.conf
 
 @call :find_erts_dir
@@ -67,7 +74,7 @@
 )
 
 :: Write the erl.ini file to set up paths relative to this script
-@call :write_ini
+:: @call :write_ini
 
 :: If a start.boot file is not present, copy one from the named .boot file
 @if not exist "%rel_dir%\start.boot" (
@@ -207,9 +214,9 @@
 @call :generate_app_config
 @set args=-detached %sys_config% %generated_config_args% -mnesia dir '%mnesia_dir%'
 @echo off
-cd /d %rel_root_dir%
 @echo on
-@start "%rel_name%" %werl% -boot "%boot_script%" %args%
+@title "%rel_name%"
+@%werl% -boot "%boot_script%" %args% || exit /b
 @goto :eof
 
 :: Stop the Windows service
@@ -217,6 +224,15 @@
 :: window service?
 :: @%erlsrv% stop %service_name%
 @%escript% %nodetool% %node_type% %node_name% -setcookie %node_cookie% stop
+@IF %ERRORLEVEL% NEQ 0 (
+    @exit /b
+)
+:TryPinging
+@%escript% %nodetool% ping %node_type% "%node_name%" -setcookie "%node_cookie%"
+@IF %ERRORLEVEL% EQU 0 (
+    @goto TryPinging
+)
+@%epmd% -kill
 @goto :eof
 
 :: Relup and reldown
@@ -237,10 +253,9 @@
 @call :generate_app_config
 @set args=%sys_config% %generated_config_args% -mnesia dir '%mnesia_dir%'
 @echo off
-cd /d %rel_root_dir%
 @echo on
-@start "bin\%rel_name% console" %werl% -boot "%boot_script%" %args%
-@echo emqx is started!
+@title "%rel_root_dir%\bin\%rel_name% console"
+@%erl_exe% -noshell -boot "%boot_script%" %args% || exit /b
 @goto :eof
 
 :: Ping the running node
@@ -255,7 +270,7 @@
 
 :: Attach to a running node
 :attach
-:: @start "%node_name% attach" 
+:: @start "%node_name% attach"
 @start "%node_name% attach" %werl% -boot "%clean_boot_script%" ^
   -remsh %node_name% %node_type% console_%node_name% -setcookie %node_cookie%
 @goto :eof
@@ -264,4 +279,3 @@
 :set_trim
 @set %1=%2
 @goto :eof
-
