--- emqx	(revision a8f6215a36953e428d902ec29d97bb2461f93e75)
+++ emqx	(date 1683920814527)
@@ -775,7 +775,7 @@
 
     ## Merge hocon generated *.args into the vm.args
     TMP_ARG_FILE="$CONFIGS_DIR/vm.args.tmp"
-    cp "$EMQX_ETC_DIR/vm.args" "$TMP_ARG_FILE"
+    cat "$EMQX_ETC_DIR/vm.args" > "$TMP_ARG_FILE"
     echo "" >> "$TMP_ARG_FILE"
     echo "-pa \"${REL_DIR}/consolidated\"" >> "$TMP_ARG_FILE"
     ## read lines from generated vm.<time>.args file
